File: 29. CheapestFlightsKStops.py
"""
Task: Cheapest Flights Within K Stops
Find the cheapest price for a flight route from a source to a destination with at most k stops. (A variation of Bellman-Ford or Dijkstra's).
"""
from collections import deque
def find_cheapest_price(n, flights, src, dst, K):
    adj = [[] for _ in range(n)]
    for u, v, w in flights:
        adj[u].append((v, w))
    dist = [float('inf')]*n
    dist[src] = 0
    q = deque([(src, 0)])
    stops = 0
    while q and stops <= K:
        sz = len(q)
        temp = dist[:]
        for _ in range(sz):
            u, cost = q.popleft()
            for v, w in adj[u]:
                if cost + w < temp[v]:
                    temp[v] = cost + w
                    q.append((v, temp[v]))
        dist = temp
        stops += 1
    return -1 if dist[dst] == float('inf') else dist[dst]
# Sample Test Case
if __name__ == "__main__":
    n, src, dst, K = 4, 0, 3, 1
    flights = [(0,1,100),(1,2,100),(2,3,100),(0,2,500)]
    print("Cheapest Price:", find_cheapest_price(n, flights, src, dst, K))
"""
Time Complexity: O(K*E)
Space Complexity: O(V)
"""

File: 29. CheapestFlightsKStops.cpp
/*
Task: Cheapest Flights Within K Stops
Find the cheapest price for a flight route from a source to a destination with at most k stops. (A
variation of Bellman-Ford or Dijkstra's).
*/
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K)
{
    vector<vector<pair<int, int>>> adj(n);
    for (auto& f : flights) adj[f[0]].push_back({f[1], f[2]});
    vector<int> dist(n, INT_MAX);
    dist[src] = 0;
    queue<pair<int, int>> q;
    q.push({src, 0});
    int stops = 0;
    while (!q.empty() && stops <= K)
    {
        int sz = q.size();
        vector<int> temp = dist;
        while (sz--)
        {
            auto [u, cost] = q.front();
            q.pop();
            for (auto& [v, w] : adj[u])
            {
                if (cost + w < temp[v])
                {
                    temp[v] = cost + w;
                    q.push({v, temp[v]});
                }
            }
        }
        dist = temp;
        stops++;
    }
    return dist[dst] == INT_MAX ? -1 : dist[dst];
}
// Sample Test Case
int main()
{
    int n = 4, src = 0, dst = 3, K = 1;
    vector<vector<int>> flights = {{0, 1, 100}, {1, 2, 100}, {2, 3, 100}, {0, 2, 500}};
    cout << "Cheapest Price: " << findCheapestPrice(n, flights, src, dst, K) << endl;
    return 0;
}
/*
Time Complexity: O(K*E)
Space Complexity: O(V)
*/
